main/src/library/Channel/Channel.flix

```diff
@@ -0,0 +1,439 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Channel {
+
+    type alias ReentrantLock = ##java.util.concurrent.locks.ReentrantLock
+
+    def newReentrantLock(): ReentrantLock & Impure =
+        import new java.util.concurrent.locks.ReentrantLock() as javaNewReentrantLock;
+        javaNewReentrantLock()
+
+    def newCondition(reentrantLock: ReentrantLock): Condition & Impure =
+        import java.util.concurrent.locks.ReentrantLock.newCondition() as javaNewCondition;
+        javaNewCondition(reentrantLock)
+
+    def java_lock(reentrantLock: ReentrantLock): Unit & Impure =
+        import java.util.concurrent.locks.ReentrantLock.lock() as javaLock;
+        javaLock(reentrantLock)
+
+    def java_unlock(reentrantLock: ReentrantLock): Unit & Impure =
+        import java.util.concurrent.locks.ReentrantLock.unlock() as javaUnlock;
+        javaUnlock(reentrantLock)
+
+    type alias Condition = ##java.util.concurrent.locks.Condition
+
+    def signalAll(cond: Condition): Unit & Impure =
+        import java.util.concurrent.locks.Condition.signalAll() as javaSignalAll;
+        javaSignalAll(cond)
+
+    def await(condition: Condition): Unit & Impure =
+        import java.util.concurrent.locks.Condition.await() as javaAwait;
+        javaAwait(condition)
+
+    /**
+    * The tuple corresponds to (id, channelLock, bufferSize, unbuffered, elementQueue, waitingGetters, waitingSetters)
+    *   - id
+    *   - channelLock is the channelLock of this channel.
+    *   - bufferSize is the size of a channel. If you try to put an element in a channel that's full, you wait until there's space.
+    *   - a flag for whether the channel is unbuffered or not
+    *   - elementQueue is the elementQueue of elements in the list.
+    *   - waitingGetters is a set of conditions that is waiting for get. This set is cleared after each new element.
+    *   - waitingSetters is a condition that can notify threads of available space in the elementQueue.
+    */
+    pub enum ChannelImpl[a] {
+        case Mpmc(
+                Int64,
+                ReentrantLock,
+                Int,
+                Bool,
+                MutList[a],
+                MutList[(ReentrantLock, Condition)],
+                Condition
+            )
+    }
+
+    /// @Override
+    /// pub def equals(o: a): Bool & Impure =
+    ///     if (this == o) return true;
+    ///     if (o == null || getClass() != o.getClass()) return false;
+    ///     ChannelIndexPair that = (ChannelIndexPair) o;
+    ///     return index == that.index && Objects.equals(channel, that.channel);
+    /// }
+
+    /// @Override
+    /// pub def hashCode(cip: ChannelIndexPair): Int =
+    ///     import java.util.Objects.hash as hash;
+    ///     hash(getChannel(cip), getIndex(cip))
+
+    /// instance Order[Channel] {
```

Comment:
It should be instance `Order[Channel[a]]` because Channel requires a type parameter.